### PR TITLE
[10] [Backend] As a user, I can fetch survey list in Home screen

### PR DESCRIPTION
**Pull request by manh-t**
https://github.com/manh-t/flutter-survey/issues/10

## What happened 👀

After logged in successfully, I can fetch survey list in Home screen.

## Proof Of Work 📹

**Request and Response**
```json
I/flutter (14888): *** Request ***
I/flutter (14888): uri: https://nimble-survey-web-staging.herokuapp.com/api/v1/surveys?page%5Bnumber%5D=1&page%5Bsize%5D=5
I/flutter (14888): method: GET
I/flutter (14888): responseType: ResponseType.json
I/flutter (14888): followRedirects: true
I/flutter (14888): connectTimeout: 3000
I/flutter (14888): sendTimeout: 0
I/flutter (14888): receiveTimeout: 5000
I/flutter (14888): receiveDataWhenStatusError: true
I/flutter (14888): extra: {}
I/flutter (14888): headers:
I/flutter (14888):  Authorization: Bearer iO7iiNDGVzFHYod-K7ZWoPgNgun66R1UNcfCQgYhc5g
I/flutter (14888): data:
I/flutter (14888): {}
I/flutter (14888): 
I/flutter (14888): *** Response ***
I/flutter (14888): uri: https://nimble-survey-web-staging.herokuapp.com/api/v1/surveys?page%5Bnumber%5D=1&page%5Bsize%5D=5
I/flutter (14888): statusCode: 200
I/flutter (14888): headers:
I/flutter (14888):  connection: keep-alive
I/flutter (14888):  cache-control: max-age=0, private, must-revalidate
I/flutter (14888):  transfer-encoding: chunked
I/flutter (14888):  date: Tue, 11 May 2021 09:28:43 GMT
I/flutter (14888):  vary: Accept-Encoding, Origin
I/flutter (14888):  content-encoding: gzip
I/flutter (14888):  strict-transport-security: max-age=31536000; includeSubDomains
I/flutter (14888):  referrer-policy: strict-origin-when-cross-origin
I/flutter (14888):  x-permitted-cross-domain-policies: none
I/flutter (14888):  content-type: application/json; charset=utf-8
I/flutter (14888):  x-xss-protection: 1; mode=block
I/flutter (14888):  server: Cowboy
I/flutter (14888):  x-request-id: 31303a42-d0ba-4587-930a-6e01a4bc5d7f
I/flutter (14888):  x-download-options: noopen
I/flutter (14888):  x-runtime: 0.035120
I/flutter (14888):  etag: W/"9b5cd44f3b9b02d18c90f6887510d818"
I/flutter (14888):  via: 1.1 vegur
I/flutter (14888):  x-frame-options: SAMEORIGIN
I/flutter (14888):  x-content-type-options: nosniff
I/flutter (14888): Response Text:
I/flutter (14888): {"data":[{"id":"d5de6a8f8f5f1cfe51bc","type":"survey","attributes":{"title":"Scarlett Bangkok","description":"We'd love ot hear from you!","thank_email_above_threshold":"<span style=\"font-family:arial,helvetica,sans-serif\"><span style=\"font-size:14px\">Dear {name},<br /><br />Thank you for visiting Scarlett Wine Bar &amp; Restaurant at Pullman Bangkok Hotel G &nbsp;and for taking the time to complete our guest feedback survey!<br /><br />Your feedback is very important to us and each survey is read individually by the management and owners shortly after it is sent. We discuss comments and suggestions at our daily meetings and use them to constantly improve our services.<br /><br />We would very much appreciate it if you could take a few more moments and review us on TripAdvisor regarding your recent visit. By <a href=\"https://www.tripadvisor.com/Restaurant_Review-g293916-d2629404-Reviews-Scarlett_Wine_Bar_Restaurant-Bangkok.html\">clicking here</a> you will be directed to our page.&nbsp;<br /><br />Thank

```

### CODE REVIEW

File: lib/api/response/base/base_http_response.dart [hunk 1]
```diff
@@ -16,3 +16,25 @@ class BaseHttpResponse<T extends BaseResponse> {
...
+      data: (json['data'] as List<dynamic>)
+          .map((e) => ApiResponse<T>.fromJson(e as Map<String, dynamic>))
+          .toList(),
+    );
+  }
+
+  Map<String, dynamic> toJson() => <String, dynamic>{
+        'data': this.data,
+      };
+}
```
Comment (markgravity) [inline lib/api/response/base/base_http_response.dart:40]: There is a trick that I also gonna try, but I'm not sure whether this is a good way.

For example:
We have a `BaseHttpResponse<T>`, then we use it like this `BaseHttpResponse<List<ApiResponse<UserResponse>>`
And the trick here that: `T.toString()` will give us `"List<ApiResponse<UserResponse>>"`, so parse this string to get `"UserResponse"`
Finally, do the rest as the normal but using string comparison rather than a type comparison

**Review by markgravity — APPROVED** [review]
LGTM 🤖

File: lib/api/response/base/base_http_response.dart (hunk 1 above)
Comment (manh-t) [reply lib/api/response/base/base_http_response.dart:40]: @markgravity good idea, let's me try it and if it sounds good, I will create a chore PR to refactor this stuff 😄

Comment (manh-t) [reply lib/api/response/base/base_http_response.dart:40]: thanks 🙏 🙏 🙏

File: .vscode/launch.json [hunk 2]
```diff
@@ -11,7 +11,8 @@
             "type": "dart",
             "program": "lib/main-staging.dart",
             "args": [
-                "--flavor=staging"
+                "--flavor",
+                "staging"
```
Comment (luongvo) [inline .vscode/launch.json:15]: Just wonder what is the purpose of this change?

File: lib/di/interceptor/app_interceptor.dart [hunk 3]
```diff
@@ -1,16 +1,18 @@
...
+  AppInterceptor(this._requireAuthenticate, this._sharedPreferenceHelper);
 
   @override
   Future onRequest(
       RequestOptions options, RequestInterceptorHandler handler) async {
     if (_requireAuthenticate) {
-      // TODO: Add Token
-      options.headers.putIfAbsent('Authorization', () => "Add Token here");
+      options.headers.putIfAbsent(
+          'Authorization', () => _sharedPreferenceHelper.authToken);
```
Comment (luongvo) [inline lib/di/interceptor/app_interceptor.dart:15]: 👍

File: .vscode/launch.json (hunk 2 above)
Comment (manh-t) [reply .vscode/launch.json:15]: @luongvo this setting help us run project on vscode with pre-defined env 🙏

**Review by luongvo — APPROVED** [review]
